src/Castle.Windsor/MicroKernel/Arguments.cs

```diff
@@ -19,208 +19,267 @@ namespace Castle.MicroKernel
 
-		public ICollection Values
-		{
-			get { return arguments.Values; }
-		}
+		bool ICollection.IsSynchronized => dictionary.IsSynchronized;
+
+		object ICollection.SyncRoot => dictionary.SyncRoot;
 
-		bool ICollection.IsSynchronized
+		bool IDictionary.IsFixedSize => dictionary.IsFixedSize;
+
+		bool IDictionary.IsReadOnly => isReadOnly;
+
+		public static readonly Arguments Empty = new Arguments() { isReadOnly = true };
+
+		/// <summary>
+		/// Method for adding named/typed dependency arguments for <see cref="IWindsorContainer.Resolve{T}(Castle.MicroKernel.Arguments)"/>
+		/// </summary>
+		/// <param named="namedOrTyped">Type or named parameter</param>
+		/// <param named="value">Instance dependency</param>
+		public void Add(object namedOrTyped, object value)
 		{
-			get { return arguments.IsSynchronized; }
+			EnsureWritable();
+			dictionary.Add(namedOrTyped, value);
 		}
 
-		object ICollection.SyncRoot
+		public void Clear()
 		{
-			get { return arguments.SyncRoot; }
+			EnsureWritable();
+			dictionary.Clear();
 		}
 
-		bool IDictionary.IsFixedSize
+		public Arguments Clone()
 		{
-			get { return arguments.IsFixedSize; }
+			if (dictionary is Dictionary<object, object>)
+			{
+				return new Arguments(dictionary);
+			}
+
+			return new Arguments(dictionary);
 		}
 
-		bool IDictionary.IsReadOnly
+		void ICollection.CopyTo(Array array, int index)
 		{
-			get { return arguments.IsReadOnly; }
+			dictionary.CopyTo(array, index);
 		}
 
-		public void Add(object key, object value)
+		public bool Contains(object key)
 		{
-			arguments.Add(key, value);
+			return dictionary.Contains(key);
 		}
 
-		public void Clear()
+		IEnumerator IEnumerable.GetEnumerator()
 		{
-			arguments.Clear();
+			return GetEnumerator();
 		}
 
-		public bool Contains(object key)
+		public IDictionaryEnumerator GetEnumerator()
 		{
-			return arguments.Contains(key);
+			return dictionary.GetEnumerator();
 		}
 
-		public IDictionaryEnumerator GetEnumerator()
+		/// <summary>
+		/// Inserts a new named argument with a given dependency. If an argument for this named key already exists, it will be overwritten.
+		/// </summary>
+		/// <param named="namedOrTyped">Named parameter</param>
+		/// <param named="value">Instance dependency</param>
+		public virtual Arguments Insert(string named, object value)
 		{
-			return arguments.GetEnumerator();
+			this[named] = value;
+			return this;
 		}
 
-		public void Remove(object key)
+		/// <summary>
+		/// Inserts dictionary with named/typed arguments. If an argument for this named/typed key already exists, it will be overwritten.
+		/// </summary>
+		/// <param named="values"></param>
+		/// <returns><see cref="Arguments"/></returns>
+		public virtual Arguments Insert(IDictionary values)
 		{
-			arguments.Remove(key);
+			foreach (DictionaryEntry entry in values)
+			{
+				if (entry.Key is string || entry.Key is Type)
+				{
+					dictionary[entry.Key] = entry.Value;
```

Comment:
Missing `EnsureWritable` since you are using `dictionary[] = x` not `this[] = x`, same applies to the following `InsertNamed` method.